Test/Unit/Controller/Cart/EmailTest.php

```diff
@@ -0,0 +1,163 @@
+        $this->customerSession = $this->createMock(CustomerSession::class);
+
+        $this->bugsnag = $this->createMock(Bugsnag::class);
+
+        $this->cartHelper = $this->createMock(CartHelper::class);
+
+        //objects needed for method
+        $this->quote = $this->createMock(Quote::class);
+        $this->quote->method('getId')->willReturn('1234');
+
+        $this->request = $this->createMock(RequestInterface::class);
+
+        //methods
+        $this->cartHelper->method('quoteResourceSave')->willReturn(null);
+        $this->context->method('getRequest')->willReturn($this->request);
+    }
+
+    private function initCurrentMock()
+    {
+        $this->currentMock = $this->getMockBuilder(Email::class)
+            ->setConstructorArgs([
+                $this->context,
+                $this->checkoutSession,
+                $this->customerSession,
+                $this->bugsnag,
+                $this->cartHelper
+            ])
+            ->enableProxyingToOriginalMethods()
+            ->getMock();
+    }
+
+    public function testExecute_quoteDoesNotExist()
+    {
+        $this->checkoutSession->method('getQuote')->willReturn(null);
+        $exception = new LocalizedException(__('Quote does not exist.'));
+        $this->bugsnag->expects($this->once())
+            ->method('notifyException')
+            ->with($this->equalTo($exception));
+        $this->currentMock->execute();
+    }
+
+    public function testExecute_noQuoteId()
+    {
+        $this->quote->method('getId')->willReturn(false);
+        $exception = new LocalizedException(__('Quote does not exist.'));
+        $this->bugsnag->expects($this->once())
+                    ->method('notifyException')
+                    ->with($this->equalTo($exception));
+        $this->currentMock->execute();
+    }
+
+    public function testExecute_noEmail()
+    {
+        $this->checkoutSession->method('getQuote')->willReturn($this->quote);
+        $this->request->method('getParam')->willReturn('');
+        $exception = new LocalizedException(__('No email received.'));
+        $this->bugsnag->expects($this->once())
+                    ->method('notifyException')
+                    ->with($this->equalTo($exception));
+        $this->currentMock->execute();
+    }
+
+    /**
+     * This doesn't really seem to test invalid emails as the validateEmail method
+     * needs to be mocked. Generally just making sure that if validateEmail comes
+     * back false we error handle properly
+     */
+    public function testExecute_invalidEmail()
+    {
+        $this->checkoutSession->method('getQuote')->willReturn($this->quote);
+        $this->request->method('getParam')->willReturn('invalidemail');
+        $this->cartHelper->method('validateEmail')->willReturn(false);
+        $exception = new LocalizedException(__('Invalid email: %1', 'invalidemail'));
+        $this->bugsnag->expects($this->once())
```

Comment:
nit: can you have these two tied to a single variable so that a change in one does not break the other?

Reply:
Done